activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -75,32 +62,24 @@ def options
         private
 
         def associated_records_by_owner(preloader)
-          owners_map = owners_by_key
-          owner_keys = owners_map.keys.compact
-
-          # Each record may have multiple owners, and vice-versa
-          records_by_owner = owners.each_with_object({}) do |owner,h|
-            h[owner] = []
-          end
-
-          if owner_keys.any?
-            # Some databases impose a limit on the number of ids in a list (in Oracle it's 1000)
-            # Make several smaller queries if necessary or make one query if the adapter supports it
-            sliced  = owner_keys.each_slice(klass.connection.in_clause_length || owner_keys.size)
-
-            records = load_slices sliced
-            records.each do |record, owner_key|
-              owners_map[owner_key].each do |owner|
-                records_by_owner[owner] << record
-              end
-            end
+          records = load_records
+          owners.each_with_object({}) do |owner, result|
+            result[owner] = records[convert_key(owner[owner_key_name])] || []
           end
+        end
 
-          records_by_owner
+        def owner_keys
+          @owner_keys ||= owners.map do |owner|
+            owner[owner_key_name]
+          end.tap {|z| z.uniq!; z.compact!}
         end
 
         def key_conversion_required?
-          association_key_type != owner_key_type
+          @key_conversion_required ||= association_key_type != owner_key_type
```

Comment:
We need that because both `*_type` calls has a `Symbol#to_s` calls that spawns new objects.